File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));
        LinkedList<int[]> list = new LinkedList<>();
        list.add(intervals[0]);
        for (int i = 1 ; i < intervals.length; i++) {
            int start = list.getLast()[0];
            int end = list.getLast()[1];
            if (end < intervals[i][0]) {
                list.add(intervals[i]);
            } else {
                list.getLast()[0] = Math.min(start, intervals[i][0]);
                list.getLast()[1] = Math.max(end, intervals[i][1]);
            }
        }
        int[][] res = new int[list.size()][2];
        int idx = 0;
        for (int[] in : list) {
            res[idx][0] = in[0];
            res[idx][1] = in[1];
            idx++;
        }
        return res;
    }
}

File: 56-merge-intervals/merge-intervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        res = []
        intervals.sort()
        for i, j in intervals:
            if not res:
                res.append([i,j])
            elif res[-1][1] < i:
                res.append([i,j])
            elif res[-1][1] >= i:
                res[-1][0] = min(res[-1][0], i)
                res[-1][1] = max(res[-1][1], j)
        return res